builder/chroots.go

```diff
@@ -80,6 +82,416 @@ func readBuildChrootsConfig(path string) (*buildChrootsConfig, error) {
+		dir := filepath.Dir(file)
+		// skip first empty string so we don't add '/'
+		for _, part := range strings.Split(dir, "/")[1:] {
+			path = filepath.Join(path, part)
+			destination[path] = true
+		}
+		destination[file] = true
+	}
+}
+
+func addOsCoreSpecialFiles(bundle *bundle) {
+	filesToAdd := []string{
+		"/usr/lib/os-release",
+		"/usr/share/clear/version",
+		"/usr/share/clear/versionstamp",
+	}
+
+	addFileAndPath(bundle.Files, filesToAdd...)
+}
+
+func addUpdateBundleSpecialFiles(b *Builder, bundle *bundle) {
+	filesToAdd := []string{
+		"/usr/share/defaults/swupd/contenturl",
+		"/usr/share/defaults/swupd/versionurl",
+		"/usr/share/defaults/swupd/format",
+	}
+
+	if _, err := os.Stat(b.Cert); err == nil {
+		filesToAdd = append(filesToAdd, "/usr/share/clear/update-ca/Swupd_Root.pem")
+	}
+
+	addFileAndPath(bundle.Files, filesToAdd...)
+}
+
+func resolveFilesForBundle(bundle *bundle, packagerCmd []string) error {
+	bundle.Files = make(map[string]bool)
+	for p := range bundle.DirectPackages {
+		queryString := merge(packagerCmd, "repoquery", "-l", "--quiet", p)
+		outBuf, err := helpers.RunCommandOutput(false, queryString[0], queryString[1:]...)
+		if err != nil {
+			return err
+		}
+		for _, f := range strings.Split(outBuf.String(), "\n") {
+			if len(f) > 0 {
+				addFileAndPath(bundle.Files, resolveFileName(f))
+			}
+		}
+	}
+	addFileAndPath(bundle.Files, fmt.Sprintf("/usr/share/clear/bundles/%s", bundle.Name))
+	fmt.Printf("Bundle %s\t%d files\n", bundle.Name, len(bundle.Files))
+	return nil
+}
+
+func resolveFiles(numWorkers int, set bundleSet, packagerCmd []string) error {
+	var err error
+	var wg sync.WaitGroup
+	fmt.Printf("Resolving files using %d workers\n", numWorkers)
+	wg.Add(numWorkers)
+	bundleCh := make(chan *bundle)
+	errorCh := make(chan error, numWorkers)
+	fileWorker := func() {
+		for bundle := range bundleCh {
+			fmt.Printf("processing %s\n", bundle.Name)
+			err = resolveFilesForBundle(bundle, packagerCmd)
+			if err != nil {
+				errorCh <- err
+			}
+		}
+		wg.Done()
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go fileWorker()
+	}
+
+	for _, bundle := range set {
+		select {
+		case bundleCh <- bundle:
+		case err = <-errorCh:
+			// Break as soon as there is a failure.
+			break
+		}
+	}
+	close(bundleCh)
+	wg.Wait()
+
+	// Sending loop might finish before any goroutine could send an error back, so check for
+	// error again after they are all done.
+	if err == nil && len(errorCh) > 0 {
+		err = <-errorCh
+	}
+	return err
+}
+
+func parseNoopInstall(installOut string) []string {
+	pkgDeps := []string{}
+	inBlock := false
+	re := regexp.MustCompile("[[:space:]]([[:graph:]]*)[[:space:]]")
+	for _, line := range strings.Split(installOut, "\n") {
+		if !inBlock {
+			if !strings.Contains(line, "Installing dependencies:") {
+				continue
+			} else {
+				inBlock = true
+			}
+		}
+
+		// at this point we are inBlock of packages list to be installed
+		// next line break is the end of the packages list
```

Comment:
I would like an example of what installOut should look like in the comments. It would make it a lot simpler to see if the code was doing the right thing, rather than seeing what the code is doing and hoping that if it is doing something that seems reasonable then it is correct. It also protects against future changes in dnf output.

Reply:
done.